File: artifacts/api-server/src/routes/drip.ts
import { Router } from "express";
import { db } from "@workspace/db";
import { dripCampaignsTable, dripStepsTable, dripEnrollmentsTable } from "@workspace/db";
import { eq, and, sql } from "drizzle-orm";
import { getUserFromToken } from "./auth";

const router = Router();

function getUser(req: any): number {
  const token = req.headers.authorization?.startsWith("Bearer ") ? req.headers.authorization.slice(7) : null;
  return (token ? getUserFromToken(token) : null) ?? 1;
}

// ── Campaigns ─────────────────────────────────────────────────────────────────

router.get("/drip", async (req, res): Promise<void> => {
  const uid = getUser(req);
  const campaigns = await db
    .select()
    .from(dripCampaignsTable)
    .where(eq(dripCampaignsTable.userId, uid))
    .orderBy(sql`${dripCampaignsTable.createdAt} DESC`);

  const result = await Promise.all(campaigns.map(async (c) => {
    const [steps] = await db
      .select({ count: sql<number>`count(*)` })
      .from(dripStepsTable)
      .where(eq(dripStepsTable.campaignId, c.id));
    const [enrolls] = await db
      .select({ count: sql<number>`count(*)` })
      .from(dripEnrollmentsTable)
      .where(and(eq(dripEnrollmentsTable.campaignId, c.id), eq(dripEnrollmentsTable.status, "active")));
    return {
      id: String(c.id), name: c.name, description: c.description,
      status: c.status, deviceId: String(c.deviceId),
      stepsCount: Number(steps?.count ?? 0),
      activeEnrollments: Number(enrolls?.count ?? 0),
      createdAt: c.createdAt?.toISOString(),
    };
  }));
  res.json(result);
});

router.post("/drip", async (req, res): Promise<void> => {
  const uid = getUser(req);
  const { name, description, deviceId, status } = req.body;
  if (!name || !deviceId) { res.status(400).json({ message: "name dan deviceId wajib diisi" }); return; }
  const [c] = await db.insert(dripCampaignsTable)
    .values({ userId: uid, deviceId: parseInt(deviceId, 10), name, description: description ?? null, status: status ?? "active" })
    .returning();
  res.json({ id: String(c.id), name: c.name, status: c.status, createdAt: c.createdAt?.toISOString() });
});

router.put("/drip/:id", async (req, res): Promise<void> => {
  const uid = getUser(req);
  const id = parseInt(req.params.id, 10);
  const { name, description, status, deviceId } = req.body;
  const [existing] = await db.select().from(dripCampaignsTable).where(and(eq(dripCampaignsTable.id, id), eq(dripCampaignsTable.userId, uid)));
  if (!existing) { res.status(404).json({ message: "Campaign tidak ditemukan" }); return; }
  const updates: any = {};
  if (name !== undefined) updates.name = name;
  if (description !== undefined) updates.description = description;
  if (status !== undefined) updates.status = status;
  if (deviceId !== undefined) updates.deviceId = parseInt(deviceId, 10);
  const [c] = await db.update(dripCampaignsTable).set(updates).where(eq(dripCampaignsTable.id, id)).returning();
  res.json({ id: String(c.id), name: c.name, status: c.status });
});

router.delete("/drip/:id", async (req, res): Promise<void> => {
  const uid = getUser(req);
  const id = parseInt(req.params.id, 10);
  const [existing] = await db.select().from(dripCampaignsTable).where(and(eq(dripCampaignsTable.id, id), eq(dripCampaignsTable.userId, uid)));
  if (!existing) { res.status(404).json({ message: "Campaign tidak ditemukan" }); return; }
  await db.delete(dripEnrollmentsTable).where(eq(dripEnrollmentsTable.campaignId, id));
  await db.delete(dripStepsTable).where(eq(dripStepsTable.campaignId, id));
  await db.delete(dripCampaignsTable).where(eq(dripCampaignsTable.id, id));
  res.json({ message: "Campaign dihapus" });
});

// ── Steps ─────────────────────────────────────────────────────────────────────

router.get("/drip/:id/steps", async (req, res): Promise<void> => {
  const uid = getUser(req);
  const id = parseInt(req.params.id, 10);
  const [campaign] = await db.select().from(dripCampaignsTable).where(and(eq(dripCampaignsTable.id, id), eq(dripCampaignsTable.userId, uid)));
  if (!campaign) { res.status(404).json({ message: "Campaign tidak ditemukan" }); return; }
  const steps = await db.select().from(dripStepsTable).where(eq(dripStepsTable.campaignId, id))
    .orderBy(dripStepsTable.stepOrder);
  res.json(steps.map((s) => ({
    id: String(s.id), stepOrder: s.stepOrder, delayDays: s.delayDays,
    message: s.message, mediaUrl: s.mediaUrl, mediaType: s.mediaType,
  })));
});

router.post("/drip/:id/steps", async (req, res): Promise<void> => {
  const uid = getUser(req);
  const id = parseInt(req.params.id, 10);
  const [campaign] = await db.select().from(dripCampaignsTable).where(and(eq(dripCampaignsTable.id, id), eq(dripCampaignsTable.userId, uid)));
  if (!campaign) { res.status(404).json({ message: "Campaign tidak ditemukan" }); return; }
  const { message, delayDays, stepOrder, mediaUrl, mediaType } = req.body;
  if (!message) { res.status(400).json({ message: "message wajib diisi" }); return; }
  const [step] = await db.insert(dripStepsTable).values({
    campaignId: id, userId: uid,
    stepOrder: stepOrder ?? 0,
    delayDays: delayDays ?? 0,
    message, mediaUrl: mediaUrl ?? null, mediaType: mediaType ?? null,
  }).returning();
  res.json({ id: String(step.id), stepOrder: step.stepOrder, delayDays: step.delayDays, message: step.message });
});

router.put("/drip/steps/:stepId", async (req, res): Promise<void> => {
  const uid = getUser(req);
  const stepId = parseInt(req.params.stepId, 10);
  const [step] = await db.select().from(dripStepsTable).where(and(eq(dripStepsTable.id, stepId), eq(dripStepsTable.userId, uid)));
  if (!step) { res.status(404).json({ message: "Step tidak ditemukan" }); return; }
  const { message, delayDays, stepOrder, mediaUrl, mediaType } = req.body;
  const updates: any = {};
  if (message !== undefined) updates.message = message;
  if (delayDays !== undefined) updates.delayDays = delayDays;
  if (stepOrder !== undefined) updates.stepOrder = stepOrder;
  if (mediaUrl !== undefined) updates.mediaUrl = mediaUrl;
  if (mediaType !== undefined) updates.mediaType = mediaType;
  const [s] = await db.update(dripStepsTable).set(updates).where(eq(dripStepsTable.id, stepId)).returning();
  res.json({ id: String(s.id), stepOrder: s.stepOrder, delayDays: s.delayDays, message: s.message });
});

router.delete("/drip/steps/:stepId", async (req, res): Promise<void> => {
  const uid = getUser(req);
  const stepId = parseInt(req.params.stepId, 10);
  const [step] = await db.select().from(dripStepsTable).where(and(eq(dripStepsTable.id, stepId), eq(dripStepsTable.userId, uid)));
  if (!step) { res.status(404).json({ message: "Step tidak ditemukan" }); return; }
  await db.delete(dripStepsTable).where(eq(dripStepsTable.id, stepId));
  res.json({ message: "Step dihapus" });
});

// ── Enrollments ───────────────────────────────────────────────────────────────

router.get("/drip/:id/enrollments", async (req, res): Promise<void> => {
  const uid = getUser(req);
  const id = parseInt(req.params.id, 10);
  const [campaign] = await db.select().from(dripCampaignsTable).where(and(eq(dripCampaignsTable.id, id), eq(dripCampaignsTable.userId, uid)));
  if (!campaign) { res.status(404).json({ message: "Campaign tidak ditemukan" }); return; }
  const enrollments = await db.select().from(dripEnrollmentsTable).where(eq(dripEnrollmentsTable.campaignId, id))
    .orderBy(sql`${dripEnrollmentsTable.enrolledAt} DESC`).limit(200);
  res.json(enrollments.map((e) => ({
    id: String(e.id), phone: e.phone, contactName: e.contactName,
    currentStep: e.currentStep, status: e.status,
    nextSendAt: e.nextSendAt?.toISOString(),
    enrolledAt: e.enrolledAt?.toISOString(),
  })));
});

router.post("/drip/:id/enroll", async (req, res): Promise<void> => {
  const uid = getUser(req);
  const id = parseInt(req.params.id, 10);
  const [campaign] = await db.select().from(dripCampaignsTable).where(and(eq(dripCampaignsTable.id, id), eq(dripCampaignsTable.userId, uid)));
  if (!campaign) { res.status(404).json({ message: "Campaign tidak ditemukan" }); return; }
  const { contacts } = req.body;
  if (!Array.isArray(contacts) || !contacts.length) { res.status(400).json({ message: "contacts wajib diisi" }); return; }
  const steps = await db.select().from(dripStepsTable).where(eq(dripStepsTable.campaignId, id)).orderBy(dripStepsTable.stepOrder);
  if (!steps.length) { res.status(400).json({ message: "Campaign belum memiliki langkah (step)" }); return; }
  const firstStep = steps[0]!;
  const nextSendAt = new Date(Date.now() + firstStep.delayDays * 86400000);
  let enrolled = 0;
  let skipped = 0;
  for (const c of contacts) {
    const phone = typeof c === "string" ? c : c.phone;
    const name = typeof c === "string" ? undefined : c.name;
    const [existing] = await db.select().from(dripEnrollmentsTable)
      .where(and(eq(dripEnrollmentsTable.campaignId, id), eq(dripEnrollmentsTable.phone, phone), eq(dripEnrollmentsTable.status, "active")));
    if (existing) { skipped++; continue; }
    await db.insert(dripEnrollmentsTable).values({
      campaignId: id, userId: uid, phone, contactName: name ?? null,
      currentStep: 0, nextSendAt, status: "active",
    });
    enrolled++;
  }
  res.json({ enrolled, skipped });
});

router.delete("/drip/enrollments/:enrollId", async (req, res): Promise<void> => {
  const uid = getUser(req);
  const enrollId = parseInt(req.params.enrollId, 10);
  const [e] = await db.select().from(dripEnrollmentsTable).where(and(eq(dripEnrollmentsTable.id, enrollId), eq(dripEnrollmentsTable.userId, uid)));
  if (!e) { res.status(404).json({ message: "Enrollment tidak ditemukan" }); return; }
  await db.update(dripEnrollmentsTable).set({ status: "cancelled" }).where(eq(dripEnrollmentsTable.id, enrollId));
  res.json({ message: "Enrollment dibatalkan" });
});

export default router;

File: go-backend/internal/handler/drip.go
package handler

import (
	"net/http"
	"strconv"
	"time"

	"github.com/Aldi1963/wagataway/internal/database/models"
	"github.com/Aldi1963/wagataway/internal/middleware"
	"github.com/gin-gonic/gin"
	"gorm.io/gorm"
)

func registerDripRoutes(rg *gin.RouterGroup, db *gorm.DB) {
	d := rg.Group("/drip")
	{
		d.GET("", listDripCampaigns(db))
		d.POST("", createDripCampaign(db))
		d.GET("/:id", getDripCampaign(db))
		d.PUT("/:id", updateDripCampaign(db))
		d.DELETE("/:id", deleteDripCampaign(db))
		d.POST("/:id/steps", addDripStep(db))
		d.PUT("/:id/steps/:stepId", updateDripStep(db))
		d.DELETE("/:id/steps/:stepId", deleteDripStep(db))
		d.POST("/:id/enroll", enrollDrip(db))
	}
}

func listDripCampaigns(db *gorm.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		userID := middleware.GetUserID(c)
		var campaigns []models.DripCampaign
		db.Where("user_id = ?", userID).Preload("Steps").
			Order("created_at DESC").Find(&campaigns)
		c.JSON(http.StatusOK, gin.H{"campaigns": campaigns})
	}
}

func createDripCampaign(db *gorm.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		userID := middleware.GetUserID(c)
		var req struct {
			Name        string `json:"name" binding:"required"`
			Description string `json:"description"`
			DeviceID    uint   `json:"deviceId" binding:"required"`
			TriggerType string `json:"triggerType"`
			TriggerVal  string `json:"triggerVal"`
		}
		if err := c.ShouldBindJSON(&req); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"message": "Data tidak valid"})
			return
		}
		if req.TriggerType == "" {
			req.TriggerType = "manual"
		}
		campaign := models.DripCampaign{
			UserID: userID, DeviceID: req.DeviceID, Name: req.Name,
			Description: req.Description, TriggerType: req.TriggerType,
			TriggerVal: req.TriggerVal, IsActive: true,
		}
		db.Create(&campaign)
		c.JSON(http.StatusCreated, gin.H{"campaign": campaign})
	}
}


func getDripCampaign(db *gorm.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		userID := middleware.GetUserID(c)
		id, _ := strconv.ParseUint(c.Param("id"), 10, 32)
		var campaign models.DripCampaign
		if err := db.Where("id = ? AND user_id = ?", id, userID).
			Preload("Steps", func(d *gorm.DB) *gorm.DB { return d.Order("step_order ASC") }).
			First(&campaign).Error; err != nil {
			c.JSON(http.StatusNotFound, gin.H{"message": "Campaign tidak ditemukan"})
			return
		}
		// Count enrollments
		var enrolled int64
		db.Model(&models.DripEnrollment{}).Where("campaign_id = ?", id).Count(&enrolled)
		c.JSON(http.StatusOK, gin.H{"campaign": campaign, "enrolled": enrolled})
	}
}

func updateDripCampaign(db *gorm.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		userID := middleware.GetUserID(c)
		id, _ := strconv.ParseUint(c.Param("id"), 10, 32)
		var campaign models.DripCampaign
		if err := db.Where("id = ? AND user_id = ?", id, userID).First(&campaign).Error; err != nil {
			c.JSON(http.StatusNotFound, gin.H{"message": "Campaign tidak ditemukan"})
			return
		}
		var req map[string]interface{}
		c.ShouldBindJSON(&req)
		db.Model(&campaign).Updates(req)
		c.JSON(http.StatusOK, gin.H{"campaign": campaign, "message": "Diperbarui"})
	}
}

func deleteDripCampaign(db *gorm.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		userID := middleware.GetUserID(c)
		id, _ := strconv.ParseUint(c.Param("id"), 10, 32)
		result := db.Where("id = ? AND user_id = ?", id, userID).Delete(&models.DripCampaign{})
		if result.RowsAffected == 0 {
			c.JSON(http.StatusNotFound, gin.H{"message": "Campaign tidak ditemukan"})
			return
		}
		// Delete steps and enrollments
		db.Where("campaign_id = ?", id).Delete(&models.DripStep{})
		db.Where("campaign_id = ?", id).Delete(&models.DripEnrollment{})
		c.JSON(http.StatusOK, gin.H{"message": "Campaign dihapus"})
	}
}

func addDripStep(db *gorm.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		userID := middleware.GetUserID(c)
		id, _ := strconv.ParseUint(c.Param("id"), 10, 32)
		var campaign models.DripCampaign
		if err := db.Where("id = ? AND user_id = ?", id, userID).First(&campaign).Error; err != nil {
			c.JSON(http.StatusNotFound, gin.H{"message": "Campaign tidak ditemukan"})
			return
		}
		var req struct {
			StepOrder  int    `json:"stepOrder"`
			DelayHours int    `json:"delayHours"`
			Type       string `json:"type"`
			Content    string `json:"content" binding:"required"`
			MediaURL   string `json:"mediaUrl"`
		}
		if err := c.ShouldBindJSON(&req); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"message": "Content wajib"})
			return
		}
		if req.Type == "" {
			req.Type = "text"
		}
		if req.DelayHours == 0 {
			req.DelayHours = 24
		}
		// Auto step order
		if req.StepOrder == 0 {
			var count int64
			db.Model(&models.DripStep{}).Where("campaign_id = ?", id).Count(&count)
			req.StepOrder = int(count) + 1
		}
		step := models.DripStep{
			CampaignID: uint(id), StepOrder: req.StepOrder,
			DelayHours: req.DelayHours, Type: req.Type,
			Content: req.Content, MediaURL: req.MediaURL,
		}
		db.Create(&step)
		c.JSON(http.StatusCreated, gin.H{"step": step})
	}
}

func updateDripStep(db *gorm.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		stepId, _ := strconv.ParseUint(c.Param("stepId"), 10, 32)
		var step models.DripStep
		if err := db.First(&step, stepId).Error; err != nil {
			c.JSON(http.StatusNotFound, gin.H{"message": "Step tidak ditemukan"})
			return
		}
		var req map[string]interface{}
		c.ShouldBindJSON(&req)
		db.Model(&step).Updates(req)
		c.JSON(http.StatusOK, gin.H{"step": step})
	}
}

func deleteDripStep(db *gorm.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		stepId, _ := strconv.ParseUint(c.Param("stepId"), 10, 32)
		db.Delete(&models.DripStep{}, stepId)
		c.JSON(http.StatusOK, gin.H{"message": "Step dihapus"})
	}
}

func enrollDrip(db *gorm.DB) gin.HandlerFunc {
	return func(c *gin.Context) {
		userID := middleware.GetUserID(c)
		id, _ := strconv.ParseUint(c.Param("id"), 10, 32)
		var campaign models.DripCampaign
		if err := db.Where("id = ? AND user_id = ?", id, userID).First(&campaign).Error; err != nil {
			c.JSON(http.StatusNotFound, gin.H{"message": "Campaign tidak ditemukan"})
			return
		}
		var req struct {
			Phones []string `json:"phones" binding:"required"`
		}
		if err := c.ShouldBindJSON(&req); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"message": "Daftar nomor wajib"})
			return
		}

		enrolled := 0
		for _, phone := range req.Phones {
			if phone == "" {
				continue
			}
			// Get first step delay
			var firstStep models.DripStep
			db.Where("campaign_id = ? AND step_order = ?", id, 1).First(&firstStep)
			delay := 24
			if firstStep.ID != 0 {
				delay = firstStep.DelayHours
			}
			nextSend := time.Now().Add(time.Duration(delay) * time.Hour)

			enrollment := models.DripEnrollment{
				CampaignID:  uint(id),
				Phone:       phone,
				CurrentStep: 0,
				Status:      "active",
				NextSendAt:  &nextSend,
				EnrolledAt:  time.Now(),
			}
			if db.Create(&enrollment).Error == nil {
				enrolled++
			}
		}

		// Update campaign enrolled count
		db.Model(&campaign).Update("enrolled", gorm.Expr("enrolled + ?", enrolled))

		c.JSON(http.StatusOK, gin.H{
			"message":  "Berhasil enroll",
			"enrolled": enrolled,
		})
	}
}
